### ISSUE TITLE
Middle Click:Make Option

### ISSUE THREAD

**Issue by fierin** [ScratchAddons/ScratchAddons#2821]
### Request

When we press middle click we can easily add blocks to the workspace, the only missing thing is creating variables, broadcasts, custom blocks, etc.

I propose that we could easily do that using the search bar. ex: 
![image](https://user-images.githubusercontent.com/86007204/122512710-da5b5a80-d03b-11eb-93b1-e4a88a123e8f.png)
which then pressing enter would create the variable and give the block.

### More examples

Variable:

![image](https://user-images.githubusercontent.com/86007204/122512785-fced7380-d03b-11eb-9870-c2973ce0b11d.png)
![image](https://user-images.githubusercontent.com/86007204/122512832-0f67ad00-d03c-11eb-8e22-2ff77978b042.png)

Additionally they COULD make a setting so instead of "var" you can use any text.

Custom Block:

![image](https://user-images.githubusercontent.com/86007204/122512961-4211a580-d03c-11eb-830c-2b59453eec90.png)
![image](https://user-images.githubusercontent.com/86007204/122512994-505fc180-d03c-11eb-8840-b9b13595f23e.png)
![image](https://user-images.githubusercontent.com/86007204/122513017-5b1a5680-d03c-11eb-859e-34d1b66ce432.png)
![image](https://user-images.githubusercontent.com/86007204/122513192-a2084c00-d03c-11eb-9699-5cbe4b0e31c0.png)

As you can see i used [ ] to define a label, ( ) to define input number or text, and < > to define input boolean.

Broadcast:

![image](https://user-images.githubusercontent.com/86007204/122513525-15aa5900-d03d-11eb-9407-b4cd7e85faed.png)
![image](https://user-images.githubusercontent.com/86007204/122513560-1f33c100-d03d-11eb-8a85-1a7b0787692e.png)

Simmilarly you could do the same for recieve blocks.

**Comment by fierin** [comment]
sorry if my request has issues. it is my first time on github

**Comment by Cool-Scratcher** [comment]
This is a really good idea. My only problem is that maybe it will be a hassle to type "var:" and then the name. If there is a project with a lot of variables that need to be made, it could get repetitive and lengthy. Maybe something else? Also, how would we specify global and local variables?

**Comment by fierin** [comment]
> This is a really good idea. My only problem is that maybe it will be a hassle to type "var:" and then the name. If there is a project with a lot of variables that need to be made, it could get repetitive and lengthy. Maybe something else? Also, how would we specify global and local variables?

actually, in most real coding languages there is a code word to type before every variable.
Also its possible to implement something like this:
1. set the code “word” in a setting.
2. have a way to mass add variables.

Also i have forgotten about lists; lists are pretty much the same as variables.

**Comment by fierin** [comment]
for lists, it really wouldnt be required to define what is in the list upon making it, in vannilla scratch this is impossible too, but we can always do something like:

list: [insert name] (insert item, insert item)

or

array: [insert name] (insert item, insert item).

**Comment by fierin** [comment]
> This is a really good idea. My only problem is that maybe it will be a hassle to type "var:" and then the name. If there is a project with a lot of variables that need to be made, it could get repetitive and lengthy. Maybe something else? Also, how would we specify global and local variables?

OH! i didnt quite see the specify global/local… well in most coding languages it would usually be only locals… we could use:

local var: ……

and

global var: …….

var: ….. 
would be the last one you did.

ofcource maybe “var:” could be changed in a setting so it could be

local // …..
etc. if you wanted to…

**Comment by fierin** [comment]
If anyone else has ideas/problems about my sujjestion, please comment so if ever this gets into the addon; it will be as refined and fool proof as possible.
Thank you Cool-Scratcher for the idea/problem!

**Comment by fierin** [comment]
quick question to the general public: how does a issue get resolved? How fast does it get resolved?

**Comment by TheColaber** [comment]
It gets resolved when someone creates a PR with changes that resolves this issue and we review it and it gets merged. Since this is not in a milestone, there is really no set date for when this will get resolved. However, you could try implementing this yourself. I don't know if anyone else is right now.

**Comment by fierin** [comment]
> It gets resolved when someone creates a PR with changes that resolves this issue and we review it and it gets merged. Since this is not in a milestone, there is really no set date for when this will get resolved. However, you could try implementing this yourself. I don't know if anyone else is right now.

Thnk you!
(though one quick thing: do PR’s require you actually fix it, or an idea how to fix it?

**Comment by fierin** [comment]
For the concern of repetitive var creating: maybe pressing shift+enter or alt+enter, etc…
would create the variable+clear the variable name+not close the middle click window.

**Comment by Secret-chest** [comment]
> > It gets resolved when someone creates a PR with changes that resolves this issue and we review it and it gets merged. Since this is not in a milestone, there is really no set date for when this will get resolved. However, you could try implementing this yourself. I don't know if anyone else is right now.
> 
> Thnk you!
> (though one quick thing: do PR’s require you actually fix it, or an idea how to fix it?

@fierin PRs require you to _fork_ this repository and make your changes to the code. Then make a PR and it will automatically detect the changed code. Then you have to wait for it to get merged and your changes will be on the main branch, which will be released in the next update.

**Comment by fierin** [comment]
thank you

**Comment by return-misaPuding** [comment]
Maybe use "var: (varname) (varvalue)" for global vars and "locvar:
(spritename) (varname) (varvalue)" (same for lists, "var" will be replaced
by "list" and items will be separated by comma tho)?

Dne út 22. 6. 2021 4:49 uživatel fierin ***@***.***> napsal:

**Comment by fierin** [comment]
> Maybe use "var: (varname) (varvalue)" for global vars and "locvar: (spritename) (varname) (varvalue)" (same for lists, "var" will be replaced by "list" and items will be separated by comma tho)? Dne út 22. 6. 2021 4:49 uživatel fierin ***@***.***> napsal:
> […](#)

indeed. but local vars would be made in the sprite you are currently on. and it may be made optional to set the value main,y because in vanilla you cant set upon creation anyway.
As i said too there should be a setting in the addon where the user can set things so maybe they could set it to:
/ [insert variable name] for global 
and
// [ins] for local if they wanted?

**Comment by fierin** [comment]
If anyone is up to the task of forking SA to make a PR on this. Please feel free.
It would amazing ti get this simetime soon.

**Comment by ErrorGamer2000** [comment]
I am currently working on an addon that will allow you to use BB code and edit the scripts as text. I think this will be a better solution.

**Comment by fierin** [comment]
Really?!?!

**Comment by fierin** [comment]
Omg ust what ive been trying to emulate.

**Comment by fierin** [comment]
though for people who love the visual scripting asthetic: it would still be usefull to have this.
Just one nquestion though: will the BB code show as visual scripting if you switch to visual?

**Comment by fierin** [comment]
> I am currently working on an addon that will allow you to use BB code and edit the scripts as text. I think this will be a better solution.

Hello! im just getting back at your BB code addon… maybe i get the linkto the issue/pull request it is being developed in?
